File: 2015/ruby/day14.rb
def assert_equal(expected, actual, str="")
  raise "Expected #{expected} but got #{actual}. #{str}" unless expected == actual
end

data = File.read('day14.txt')
def parse(line)
  match = /(?<rate>\d+).*for (?<t1>\d+) seconds.*rest for (?<t2>\d+) seconds/.match(line)
  raise "Cannot parse #{line}" if match.nil?
  match.named_captures
end
class Reindeer
  def initialize(hsh)
    @rate = hsh['rate'].to_i
    @t1 = hsh['t1'].to_i
    @t2 = hsh['t2'].to_i
  end

  def period
    @t1 + @t2
  end

  def distance(t)
    full_periods = t / period
    partial_period = t % period
    (full_periods * @t1 + [partial_period, @t1].min) * @rate
  end
end

def race(input, t)
  reindeer = input.lines.map {|l| Reindeer.new(parse(l))}
  reindeer.map{|r|r.distance(t)}.max
end
sample = <<~TXT
Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.
Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds.
TXT
assert_equal(1120, Reindeer.new(parse("Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.")).distance(1_000))
assert_equal(1056, Reindeer.new(parse("Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds.")).distance(1_000))
assert_equal(1120, race(sample, 1_000))
puts "Part 1: #{race(data, 2503)}"

File: 2015/go/day14/day14.go
package day14

import "fmt"

type Reindeer struct {
	name string
	rate int
	t1   int
	t2   int
}

func Parse(line string) Reindeer {
	var name string
	var rate, t1, t2 int
	fmt.Sscanf(line, "%s can fly %d km/s for %d seconds, but then must rest for %d seconds.", &name, &rate, &t1, &t2)
	return Reindeer{name, rate, t1, t2}
}

func Race(input []string, time int) int {
	max := 0
	for _, line := range input {
		reindeer := Parse(line)
		d := Distance(reindeer, time)
		if d > max {
			max = d
		}
	}
	return max
}
func Points(input []string, time int) int {
	scores := make([]int, len(input))
	reindeer := make([]Reindeer, len(input))
	for idx, line := range input {
		reindeer[idx] = Parse(line)
	}
	for i := 1; i <= time; i++ {
		currentDistances := make([]int, len(reindeer))
		for ri := 0; ri < len(reindeer); ri++ {
			currentDistances[ri] = Distance(reindeer[ri], i)
		}
		_, maxD, _ := MinMax(currentDistances)

		for idx, distance := range currentDistances {
			if distance == maxD {
				scores[idx] += 1
			}
		}
	}
	_, max, _ := MinMax(scores)
	return max
}
func MinMax(vals []int) (int, int, bool) {
	if len(vals) == 0 {
		return 0, 0, false
	}
	min := vals[0]
	max := vals[0]
	for _, v := range vals {
		if min > v {
			min = v
		}
		if max < v {
			max = v
		}
	}
	return min, max, true
}

func Distance(reindeer Reindeer, time int) int {
	period := reindeer.t1 + reindeer.t2
	fullPeriods := time / period
	partialPeriod := time % period
	minVal, _, _ := MinMax([]int{partialPeriod, reindeer.t1})
	return (fullPeriods*reindeer.t1 + minVal) * reindeer.rate
}
